File: app/src/main/java/com/example/merchio/LoginActivity.java
package com.example.merchio;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Patterns;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.merchio.db.DbHelper;

public class LoginActivity extends AppCompatActivity {

    EditText edtEmailLogin, edtPasswordLogin;
    Button btnLogin;
    TextView txtGoRegister, txtForgotPassword;

    DbHelper dbHelper;
    SessionManager sessionManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        dbHelper = new DbHelper(this);
        dbHelper.getWritableDatabase();
        sessionManager = new SessionManager(this);

        dbHelper.ensureRoleColumnExists();
        dbHelper.createDefaultAdminIfNeeded();

        if (sessionManager.isLoggedIn()) {
            int savedUserId = sessionManager.getUserId();
            String role = dbHelper.getUserRole(savedUserId);

            Intent intent;

            if ("admin".equalsIgnoreCase(role)) {
                intent = new Intent(LoginActivity.this, AdminDashboardActivity.class);
            } else {
                intent = new Intent(LoginActivity.this, MainActivity.class);
            }

            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            finish();
            return;
        }

        edtEmailLogin = findViewById(R.id.edtEmailLogin);
        edtPasswordLogin = findViewById(R.id.edtPasswordLogin);
        btnLogin = findViewById(R.id.btnLogin);
        txtGoRegister = findViewById(R.id.txtGoRegister);
        txtForgotPassword = findViewById(R.id.txtForgotPassword);

        btnLogin.setOnClickListener(v -> loginUser());

        txtGoRegister.setOnClickListener(v -> {
            startActivity(new Intent(LoginActivity.this, RegisterActivity.class));
            finish();
        });

        txtForgotPassword.setOnClickListener(v -> {
            Toast.makeText(this, "Fitur forgot password belum dibuat", Toast.LENGTH_SHORT).show();
        });
    }

    private void loginUser() {
        String email = edtEmailLogin.getText().toString().trim();
        String password = edtPasswordLogin.getText().toString().trim();

        if (TextUtils.isEmpty(email)) {
            edtEmailLogin.setError("Email wajib diisi");
            edtEmailLogin.requestFocus();
            return;
        }

        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            edtEmailLogin.setError("Format email tidak valid");
            edtEmailLogin.requestFocus();
            return;
        }

        if (TextUtils.isEmpty(password)) {
            edtPasswordLogin.setError("Password wajib diisi");
            edtPasswordLogin.requestFocus();
            return;
        }

        if (password.length() < 8) {
            edtPasswordLogin.setError("Password minimal 8 karakter");
            edtPasswordLogin.requestFocus();
            return;
        }

        int userId = dbHelper.loginUser(email, password);

        if (userId != -1) {
            sessionManager.saveLogin(userId, email);

            Toast.makeText(this, "Login berhasil", Toast.LENGTH_SHORT).show();

            String role = dbHelper.getUserRole(userId);

            Intent intent;

            if ("admin".equalsIgnoreCase(role)) {
                intent = new Intent(LoginActivity.this, AdminDashboardActivity.class);
            } else {
                intent = new Intent(LoginActivity.this, MainActivity.class);
            }

            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            finish();
        } else {
            Toast.makeText(this, "Email atau password salah", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/merchio/LoginActivity.kt
package com.example.merchio

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.merchio.data.repository.AuthRepository
import com.example.merchio.utils.textString
import com.example.merchio.utils.toast
import android.widget.EditText
import kotlinx.coroutines.launch

class LoginActivity : AppCompatActivity() {
    private val authRepository = AuthRepository()
    private lateinit var emailInput: EditText
    private lateinit var passwordInput: EditText
    private lateinit var loginButton: View
    private lateinit var registerText: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        emailInput = findViewById(R.id.edtEmailLogin)
        passwordInput = findViewById(R.id.edtPasswordLogin)
        loginButton = findViewById(R.id.btnLogin)
        registerText = findViewById(R.id.txtGoRegister)

        registerText.setOnClickListener { startActivity(Intent(this, RegisterActivity::class.java)) }
        loginButton.setOnClickListener { login() }
    }

    private fun login() {
        val email = emailInput.textString()
        val password = passwordInput.textString()
        if (email.isBlank()) { emailInput.error = "Email wajib diisi"; return }
        if (password.isBlank()) { passwordInput.error = "Password wajib diisi"; return }

        loginButton.isEnabled = false
        lifecycleScope.launch {
            runCatching { authRepository.signIn(email, password) }
                .onSuccess { profile ->
                    toast("Login berhasil")
                    val target = if (profile.role == "admin") AdminDashboardActivity::class.java else MainActivity::class.java
                    startActivity(Intent(this@LoginActivity, target))
                    finishAffinity()
                }
                .onFailure { toast(it.message ?: "Login gagal") }
            loginButton.isEnabled = true
        }
    }
}
